Fix #32 try to use findPnpApi to locate pnpapi

missing1984 opened the pull request:
## Context
The current implementation assume `pnp-webpack-plugin` live within the current pnp runtime. It's not true when webpack live outside of the pnp runtime and used to build pnp projects. 

It's a rare cases but Cypress uses this pattern to built its plugin.  Make pnp-webpack-plugin work with other pnp runtime is the first step to unblock yarn2 support in Cypress.

## Solution
Switch to use `findPnpApi` to locate pnp api first and fallback to require('pnpapi') if `findPnpApi` failed to locate the api.

## Effects
Fixes #32 

The fix should be compatible with no breaking change

arcanis commented on index.js:
```diff
@@ -36,7 +60,9 @@ function getSourceLocation(sourceLocator) {
     return null;
 
-  const pnp = require(`pnpapi`);
+  const pnp = getPnpApi();
+  if (!pnp) 
+    return null;
```
Doesn't that risk being a problem, if it just always return the main process' API?

arcanis commented:
I feel like this approach is a bit risky (for your use case), because different functions may end up using different instances of the API. Instead, I think an ideal approach might be to retrieve the right PnP API when instantiating the plugin, and passing it around as parameter (instead of them relying on the global `pnpapi` variable). Does that make sense?

missing1984 commented:
that works too but we need to refactor the apis which will be a breaking change..

so basically.

from 
```
function makeResolver(sourceLocator, filter) {
  const sourceLocation = getSourceLocation(sourceLocator);
 const pnpApi = require('pnpapi')
```
to something like
```
function makeResolver(sourceLocator, filter, opts) {
    const sourceLocation = getSourceLocation(sourceLocator, opts);
    const api = opts.pnpApi? opts.pnpApi: require('pnpapi'); // or take a path that can be used to fetch pnpapi.
}
```
the default export as well as other exports that immediately get invoked needs to be change to a function so consumer can pass the pnpapi or path when initialize the plugin.

does that sounds right?

missing1984 commented:
I've refactored the code a bit.

1.  Move `makeResolver` function to a `./resolver` file and refactored its interface to accept a pnpapi instance.
2. Refactor all exports from index.js to use the new `makeResolver` function with a resolved pnpapi instance.

There will be no breaking change to the existing user.. and i can use the `makeResolver` function to construct a new Webpack plugin for my use case.. for example.

```typescript
/**
 * Config yarn pnp plugin for webpack 4
 * @param {*} file file to be processed
 * @param {*} options 
 */
const addYarnPnpConfig = (file, options) => {
  const { makeResolver } = require('pnp-webpack-plugin/resolver');
  const findPnpApi = require('module').findPnpApi;
  if (findPnpApi && file.filePath) {
    const pnpapi = findPnpApi(file.filePath);
    console.log(`addYarnPnpConfig pnpapi`, pnpapi);
    const PnpPlugin = {
      apply: makeResolver({pnpapi})
    }
    options.webpackOptions.resolve.plugins.push(PnpPlugin);
  }
}

```
The tradeoff would be a new `pnp-webpack-plugin/resolver` module that we need to maintain.  I think it should be fine as we're moving to webpack 5 anyway.

missing1984 commented:
@arcanis any updates on this?
